kips/kip-2.md

```diff
@@ -0,0 +1,462 @@
+
+Storing all ETH into a single contract may seem risky, but we can negate this with good testing and auditing of the contract, and by managing admin access using multisig accounts with op-sec supervision.
+
+## Implementation
+
+_Note: the code below does not take into account ERC-20 funded events, just ETH ones_
+
+**EternalStorage**
+
+The base class which specifies an extendable storage system that allows for easy upgrades.
+
+```solidity
+/**
+ * Base storage class.
+ */
+contract EternalStorage {
+  // scalars
+  mapping(string => address) dataAddress;
+  mapping(string => string) dataString;
+  mapping(string => bytes32) dataBytes32;
+  mapping(string => int256) dataInt256;
+  mapping(string => uint256) dataUint256;
+  mapping(string => bool) dataBool;
+  // arrays
+  mapping(string => address[]) dataManyAddresses;
+  mapping(string => bytes32[]) dataManyBytes32s;
+  mapping(string => int256[]) dataManyInt256;
+  mapping(string => uint256[]) dataManyUint256;
+  mapping(string => bool[]) dataManyBool;
+}
+```
+
+**Proxy**
+
+Base class for all upgradeable contracts. Upgradeability is based on the [OpenZepellin Proxy + EternalStorage](https://github.com/OpenZeppelin/openzeppelin-labs/tree/master/upgradeability_using_eternal_storage) pattern.
+
+```solidity
+/**
+ * Upgradeable contracts base class
+ */
+contract Proxy is EternalStorage {
+  /**
+  * @dev This event will be emitted every time the implementation gets upgraded
+  * @param implementation representing the address of the upgraded implementation
+  */
+  event Upgraded(address indexed implementation);
+
+  /**
+   * Constructor.
+   */
+  constructor (address _implementation) public {
+    require(_implementation != address(0), 'implementation must be valid');
+    dataAddress["implementation"] = _implementation;
+  }
+
+  /**
+  * @dev Get the address of the implementation where every call will be delegated.
+  * @return address of the implementation to which it will be delegated
+  */
+  function getImplementation() public view returns (address) {
+    return dataAddress["implementation"];
+  }
+
+  /**
+   * @dev Point to a new implementation.
+   * This is internal so that descendants can control access to this in custom ways.
+   */
+  function setImplementation(address _implementation) internal {
+    require(_implementation != address(0), 'implementation must be valid');
+    require(_implementation != dataAddress["implementation"], 'already this implementation');
+
+    dataAddress["implementation"] = _implementation;
+
+    emit Upgraded(_implementation);
+  }
+
+  /**
+  * @dev Fallback function allowing to perform a delegatecall to the given implementation.
+  * This function will return whatever the implementation call returns
+  */
+  function () payable external {
+    address _impl = getImplementation();
+    require(_impl != address(0), 'implementation not set');
+
+    assembly {
```

Comment:
Are they exact replica of https://github.com/OpenZeppelin/openzeppelin-labs/blob/master/upgradeability_using_eternal_storage/contracts/Proxy.sol ? If so, better to import their library when we actually implement these?

Reply:
As long as we can use external library of Proxy and Eternal Storage which have been battle tested, I guess we can almost treat it like black box though delegatecall and bunch of assembly gives me a bit of chill in general.

Reply:
Get over the fear.